src/psyclone/gocean1p0.py

```diff
@@ -947,74 +948,122 @@ def __init__(self, call, parent=None):
+                                offset = 0
+
+                            inner_index = Reference(inner_symbol)
+                            outer_index = Reference(outer_symbol)
+                            if j > 0:
+                                inner_index = BinaryOperation.create(
+                                    BinaryOperation.Operator.ADD,
+                                    inner_index,
+                                    Literal(str(abs(offset)), INTEGER_TYPE))
+                            elif j < 0:
+                                inner_index = BinaryOperation.create(
+                                    BinaryOperation.Operator.SUB,
+                                    inner_index,
+                                    Literal(str(abs(offset)), INTEGER_TYPE))
+
+                            if i > 0:
+                                outer_index = BinaryOperation.create(
+                                    BinaryOperation.Operator.ADD,
+                                    outer_index,
+                                    Literal(str(abs(offset)), INTEGER_TYPE))
+                            elif i < 0:
+                                outer_index = BinaryOperation.create(
+                                    BinaryOperation.Operator.SUB,
+                                    outer_index,
+                                    Literal(str(abs(offset)), INTEGER_TYPE))
+
+                            access = StructureReference.create(symbol, [
+                                ("data", [inner_index, outer_index])
+                            ])
+                            if arg.access in (AccessType.WRITE,
+                                              AccessType.READWRITE):
+                                # if write_access is not None:
+                                #     raise InternalError(
+                                #         "This is not a valid kernel, a "
+                                #         "kernel can only write to one field"
+                                #     )
+                                write_access = access.copy()
+                            if arg.access != AccessType.WRITE:
+                                read_accesses.append(access)
+
+        # Now create the assignment prototype
+        # if not write_access:
+        #     raise InternalError(
+        #         "This is not a valid kernel, a kernel must write "
+        #         "to one field.")
+        if not read_accesses:
+            # There can be write-only kernels, in this case we just put
+            # a constant literal
+            rhs = Literal("1", INTEGER_TYPE)
+        else:
+            rhs = read_accesses.pop(0)
+            while read_accesses:
+                rhs = BinaryOperation.create(
+                        BinaryOperation.Operator.ADD,
+                        rhs, read_accesses.pop(0))
+        return Assignment.create(write_access, rhs)
```

Comment:
I do realise a small change here, which might have huge impact. As far as I can see, now all psydata libraries will not get the field provided anymore, but the `field%data` 2D Fortran array.

Now, the good news of this is of course it might mean we don't need domain-specific implementations for gocean anymore.

The downside - it's hard to evaluate if we might break something somewhere :(  That statement is triggered by the fact that only very few changes are required in the tests. Looking at `readonly`, indeed, the only change would be the error message (not talking of GOcean 'field' anymore, instead it would be array). The gocean example passes. I will have to think about this a bit, and look into other psy data libraries.